runner.py

```diff
@@ -0,0 +1,134 @@
+
+run_func_ltable = {
+    SIM_TYPE_FED_LEARNING: RunFuncAndReqParams(
+        runner_run_funcs.run_fed_learn_sim,
+        {
+            P_KEY_NUM_ROUNDS,
+            P_KEY_BATCH_SIZE,
+            P_KEY_NUM_EPOCHS,
+            P_KEY_NUM_SAMPLES,
+            P_KEY_NUM_FEATURES,
+            P_KEY_NUM_USERS,
+        },
+    ),
+    SIM_TYPE_FED_AVG_WITH_DP: RunFuncAndReqParams(
+        runner_run_funcs.run_fed_avg_with_dp, {P_KEY_NUM_LABELS, P_KEY_NUM_FEATURES}
+    ),
+    DATA_GEN_TYPE_DATA_FROM_FILE: RunFuncAndReqParams(
+        runner_run_funcs.read_data_from_file,
+        {
+            P_KEY_NUM_SAMPLES,
+            P_KEY_NUM_LABELS,
+            P_KEY_NUM_FEATURES,
+            P_KEY_NUM_USERS,
+            P_KEY_DATA_FILE_PATH,
+        },
+    ),
+    DATA_GEN_TYPE_BLOB: RunFuncAndReqParams(
+        runner_run_funcs.run_data_gen_blob,
+        {P_KEY_NUM_SAMPLES, P_KEY_NUM_LABELS, P_KEY_NUM_FEATURES, P_KEY_NUM_USERS},
+    ),
+    DATA_GEN_TYPE_RAND: RunFuncAndReqParams(
+        runner_run_funcs.run_data_gen_rand,
+        {P_KEY_NUM_SAMPLES, P_KEY_NUM_LABELS, P_KEY_NUM_FEATURES, P_KEY_NUM_USERS},
+    ),
+}
+
+
+class Runner:
+    def __init__(self):
+        self._params = {}
+
+    def run(self, sim_type, data_gen_type):
+        sim_run_info = run_func_ltable[sim_type]
+        data_gen_run_info = run_func_ltable[data_gen_type]
+
+        if not self._sim_has_required_params_for_given_run_func(
+            sim_run_info.prereq_params, sim_type
+        ):
+            return
+
+        if not self._sim_has_required_params_for_given_run_func(
+            sim_run_info.prereq_params, data_gen_type
+        ):
+            return
+
+        print('Generating "{}" data...'.format(data_gen_type))
+        generated_data = data_gen_run_info.run_func(self._params)
+
+        print('Running the "{}" simulation...'.format(sim_type))
+        sim_run_info.run_func(self._params, generated_data)
+
+        print("Finished!")
+
+    def set_num_samples(self, num_samples):
+        return self._set_param(P_KEY_NUM_SAMPLES, num_samples)
+
+    def set_num_labels(self, num_labels):
+        return self._set_param(P_KEY_NUM_LABELS, num_labels)
+
+    def set_num_features(self, num_features):
+        return self._set_param(P_KEY_NUM_FEATURES, num_features)
+
+    def set_num_users(self, num_users):
+        return self._set_param(P_KEY_NUM_USERS, num_users)
+
+    def set_read_data_file_path(self, file_path):
+        return self._set_param(P_KEY_DATA_FILE_PATH, file_path)
+
+    def set_num_rounds(self, num_rounds):
+        return self._set_param(P_KEY_NUM_ROUNDS, num_rounds)
+
+    def set_batch_size(self, batch_size):
+        return self._set_param(P_KEY_BATCH_SIZE, batch_size)
+
+    def set_num_epochs(self, num_epochs):
+        return self._set_param(P_KEY_NUM_EPOCHS, num_epochs)
+
+    def _set_param(self, p_key, val):
+        self._params[p_key] = val
+        return self
+
+    def _sim_has_required_params_for_given_run_func(
```

Comment:
Let's call this from the Runner constructor so that we can't ever have a Runner in an invalid startup state.

Reply:
Sure.

Reply:
`_sim_has_required_params_for_given_run_func` should probably throw an exception if invalid parameters have been passed in.